README.md

```diff
@@ -1,20 +1,506 @@
+
+You might wonder why this is an improvement. There are a couple of reasons, most
+importantly that modeling the above problem with a hash table is much more
+intuitive considering the most important property of the above data structure:
+
+Objects have unique keys. You "query" a hash table by getting the value
+associated with a key. Objects are kind of similar to arrays in that aspect:
+Array items can be accessed by index (e.g. `connections[1]` gives you the
+second connection), objects can be accessed by key (e.g.
+`connections['max']` or `connections.max`). We're going to go into more detail
+on how do access object properties later.
+
+The ability to get object values by key in O(1) (constant time), solves **all**
+problems we encountered when storing connections in a massive array of pairs.
+
+Checking if a username is already taken is easy: We just need to check if there
+is a socket associated with the username in question.
+
+Registering and removing client connections is trivial, since we no longer need
+to iterate over the entire array. Objects provide special means for assigning
+and removing key-value pairs.
+
+## Creating an object
+So this was a lot of theory, but how do we **actually** store the client
+connections in JavaScript? Talk is cheap, show me the code!
+
+There are three ways to create objects in JavaScript, let's look at the simplest
+one, the so-called object "literal". An object literal is simply a way to
+declare an object in its entirety, including all its keys and values, "upfront".
+
+### Object literals
+Let's look at the simplest case first, an empty object:
+
+```js
+const connections = {};
+```
+
+This isn't too useful yet. Let's say we want to "bootstrap" our server with a
+couple of client connections or want to reserve a couple of hard-coded
+usernames:
+
+```js
+const connections = {
+    max:    conn0,
+    alex:   conn1,
+    james:  conn2
+};
+```
+
+Doesn't look too different from the above table, does it? Keys are on the left,
+values on the right.
+
+Since object keys are **always** "stringified" (meaning that keys are always
+strings, never numbers or anything else) in JavaScript, we can leave out the
+enclosing `'` or `"`. Nevertheless, leaving them in also works without any
+problems:
+
+```js
+const connections = {
+    "max":  conn0, // enclosing "
+    'alex': conn1, // enclosing '
+    james:  conn2
+};
+```
+
+An interesting scenario is when you want to be able to initialize an object
+literal with something that hasn't been evaluated yet, e.g. let's say you want
+to create an object with the current date as a key. Thankfully ES6 defines an
+new syntax to achieve this:
+
+```js
+{ [Date.now()]: null }; // #=> { '1469397041282': null }
+```
+
+Keep in mind that keys are **always** strings, never arrays. While in the above
+example we enclose the key in brackets, this does not indicate that the key
+*itself* is an array.
+
+### Constructor functions
```

Comment:
My worry with this section is that students won't yet have made the connection that function is an object. It might be better to ease them into this concept gradually over a few readings and labs.